Why did not the validation build for #9305 fail? The ChangeHierarchyBeforeResponse test seems to produce an UI roughly 50% of the time and a stack trace 50% of the time:
```
javax.servlet.ServletException: com.vaadin.server.ServiceException: java.lang.AssertionError: Connector should be registered with this ConnectorTracker
	at com.vaadin.server.VaadinServlet.service(VaadinServlet.java:383)
	at com.vaadin.launcher.ApplicationRunnerServlet.service(ApplicationRunnerServlet.java:160)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:808)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:497)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.vaadin.server.ServiceException: java.lang.AssertionError: Connector should be registered with this ConnectorTracker
	at com.vaadin.server.VaadinService.handleExceptionDuringRequest(VaadinService.java:1578)
	at com.vaadin.server.VaadinService.handleRequest(VaadinService.java:1538)
	at com.vaadin.server.VaadinServlet.service(VaadinServlet.java:381)
	... 20 more
Caused by: java.lang.AssertionError: Connector should be registered with this ConnectorTracker
	at com.vaadin.ui.ConnectorTracker.isClientSideInitialized(ConnectorTracker.java:211)
	at com.vaadin.server.communication.UidlWriter.write(UidlWriter.java:107)
	at com.vaadin.server.communication.UIInitHandler.getInitialUidl(UIInitHandler.java:291)
	at com.vaadin.server.communication.UIInitHandler.synchronizedHandleRequest(UIInitHandler.java:80)
	at com.vaadin.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:41)
	at com.vaadin.server.VaadinService.handleRequest(VaadinService.java:1526)
	... 21 more
```

--- a/server/src/main/java/com/vaadin/server/communication/UidlWriter.java
+++ b/server/src/main/java/com/vaadin/server/communication/UidlWriter.java
@@ -90,10 +90,9 @@ public class UidlWriter implements Serializable {
 
         while (true) {
             ArrayList<ClientConnector> connectorsToProcess = new ArrayList<>();
-            for (ClientConnector c : uiConnectorTracker.getDirtyConnectors()) {
-                if (!processedConnectors.contains(c)
-                        && LegacyCommunicationManager
-                                .isConnectorVisibleToClient(c)) {
+            for (ClientConnector c : uiConnectorTracker
+                    .getDirtyVisibleConnectors()) {
+                if (!processedConnectors.contains(c)) {
                     connectorsToProcess.add(c);
                 }
             }
@@ -102,9 +101,25 @@ public class UidlWriter implements Serializable {
                 break;
             }
 
+            // process parents before children
+            Collections.sort(connectorsToProcess,
+                    Comparator.comparingInt(conn -> {
+                        int depth = 0;
+                        ClientConnector connector = conn;
+                        // this is a very fast operation, even for 100+ levels
+                        while (connector.getParent() != null) {
+                            ++depth;
+                            connector = connector.getParent();
+                        }
+                        return depth;
+                    }));
+
             for (ClientConnector connector : connectorsToProcess) {
-                boolean initialized = uiConnectorTracker
-                        .isClientSideInitialized(connector);
+                // call isDirty() to find out if ConnectorTracker knows the
+                // connector
+                boolean initialized = uiConnectorTracker.isDirty(connector)
+                        && uiConnectorTracker
+                                .isClientSideInitialized(connector);
                 processedConnectors.add(connector);
 
                 try {

